contracts/governance/Vesting/fouryear/FourYearVestingLogic.sol

```diff
@@ -126,6 +127,42 @@ contract FourYearVestingLogic is IFourYearVesting, FourYearVestingStorage, Appro
 		emit DividendsCollected(msg.sender, _loanPoolToken, _receiver, _maxCheckpoints);
 	}
 
+	/**
+	 * @notice Change token owner - only vesting owner is allowed to change.
+	 * @dev Modifies token owner. This must be followed by approval
+	 * from token owner.
+	 * @param _newTokenOwner Address of new token owner.
+	 * */
+	function changeTokenOwner(address _newTokenOwner) public onlyOwner {
+		require(_newTokenOwner != address(0), "invalid new token owner address");
+		require(_newTokenOwner != tokenOwner, "same owner not allowed");
+		newTokenOwner = _newTokenOwner;
+	}
+
+	/**
+	 * @notice Approve token owner change - only token Owner.
+	 * @dev Token owner can only be modified
+	 * when both vesting owner and token owner have approved. This
+	 * function ascertains the approval of token owner.
+	 * */
+	function approveOwnershipTransfer() public onlyTokenOwner {
+		require(newTokenOwner != address(0), "invalid address");
+		tokenOwner = newTokenOwner;
+		newTokenOwner = address(0);
+		emit TokenOwnerChanged(tokenOwner, msg.sender);
+	}
+
+	/**
+	 * @notice Set address of the implementation - only Token Owner.
+	 * @dev This function sets the new implementation address.
+	 * It must also be approved by the Vesting owner.
+	 * @param _newImplementation Address of the new implementation.
+	 * */
+	function setImpl(address _newImplementation) public onlyTokenOwner {
+		require(_newImplementation != address(0), "invalid new implementation address");
```

Comment:
```suggestion
		require(Address.isContract(_newImplementation), "invalid new implementation address");
```

Reply:
I have kept a check here: https://github.com/DistributedCollective/Sovryn-smart-contracts/blob/1ae486d26377a2e25e2c81e4d4469f7cc2227dea/contracts/governance/Vesting/fouryear/FourYearVesting.sol#L63 We probably don't need this.

Reply:
i think we can then remove verification completely from here as it will be done in the proxy anyway and it is not upgradeable so we are on the safe side

Reply:
But we check for the address that is set here and hence put the zero address check. I haven't put the zero address check on the proxy.